zebra-state/src/service/block_iter.rs

```diff
@@ -1,102 +1,86 @@
-    Finalized(block::Height),
-    Finished,
+#[derive(Clone, Debug)]
+pub(crate) struct Iter<Item: ChainItem> {
+    /// The non-finalized chain fork we're iterating, if the iterator is in the non-finalized state.
+    ///
+    /// This is a cloned copy of a potentially out-of-date chain fork.
+    pub(super) chain: Option<Arc<Chain>>,
+
+    /// The finalized database we're iterating.
+    ///
+    /// This is the shared live database instance, which can concurrently write blocks.
+    pub(super) db: ZebraDb,
+
+    /// The height of the item which will be yielded by `Iterator::next()`.
+    pub(super) height: Option<Height>,
+
+    /// An internal marker type that tells the Rust type system what we're iterating.
+    iterable: PhantomData<Item::Type>,
 }
 
-impl Iter<'_> {
-    fn next_non_finalized_block(&mut self) -> Option<Arc<Block>> {
-        let Iter {
-            non_finalized_state,
-            db: _,
-            state,
-        } = self;
-
-        let hash = match state {
-            IterState::NonFinalized(hash) => *hash,
-            IterState::Finalized(_) | IterState::Finished => unreachable!(),
-        };
-
-        if let Some(block) = non_finalized_state.any_block_by_hash(hash) {
-            let hash = block.header.previous_block_hash;
-            self.state = IterState::NonFinalized(hash);
-            Some(block)
-        } else {
-            None
-        }
-    }
-
-    #[allow(clippy::unwrap_in_result)]
-    fn next_finalized_block(&mut self) -> Option<Arc<Block>> {
-        let Iter {
-            non_finalized_state: _,
-            db,
-            state,
-        } = self;
-
-        let hash_or_height: HashOrHeight = match *state {
-            IterState::Finalized(height) => height.into(),
-            IterState::NonFinalized(hash) => hash.into(),
-            IterState::Finished => unreachable!(),
-        };
-
-        if let Some(block) = db.block(hash_or_height) {
-            let height = block
-                .coinbase_height()
-                .expect("valid blocks have a coinbase height");
-
-            if let Some(next_height) = height - 1 {
-                self.state = IterState::Finalized(next_height);
+impl<Item> Iter<Item>
+where
+    Item: ChainItem,
+{
+    /// Returns an item by height, and updates the iterator's internal state to point to the
+    /// previous height.
+    fn yield_by_height(&mut self) -> Option<Item::Type> {
+        let current_height = self.height?;
+
+        // TODO:
+        // Check if the root of the chain connects to the finalized state. Cloned chains can become
+        // disconnected if they are concurrently pruned by a finalized block from another chain
+        // fork. If that happens, the iterator is invalid and should stop returning items.
+        //
+        // Currently, we skip from the disconnected chain root to the previous height in the
+        // finalized state, which is usually ok, but could cause consensus or light wallet bugs.
```

Comment:
Optional: Do we want to follow up with a copy of this that iterates by `prev_hash`?

Reply:
If it is ever needed, or if this bug ever happens, we should change to iterating by hash then.

I don't think we should have copies of code that do the same thing, but one of them has a concurrency bug.